### ISSUE TITLE
Plotting the response of categorical variables

### ISSUE THREAD

**Issue by luispedrosantiago** [biomodhub/biomod2#61]
Hi Maya,

In the previous version of biomod2 it was possible to plot the response of the occurrence probability when it comes to categorical variables. However, I am unable to do so now and I am prompted with this message: "Using ForestCover10_1000m as id variables
geom_path: Each group consists of only one observation. Do you need to adjust the group
aesthetic?"

This is a rough outline of my code:

ggplot(bm_PlotResponseCurves(...)) + geom_line + geom_point

Thanks in advance,
Luís

**Comment by MayaGueguen** [comment]
Hello Luis,

The `bm_PlotResponseCurves` function should provide you directly with a `ggplot` object, you don't need to put it into a ggplot command.
However, the function is supposed to be able to deal with factorial variables.
Could you provide me please your `BIOMOD_Modeling` object (the `bm.out` parameter of the function), as well as your explanatory dataset that you gave to the `BIOMOD_FormatingData` ?
Thanks in advance

Maya

**Comment by DelphineChabanne** [comment]
Hi Maya,

I am having the same issue, trying to plot the variables. One is categorical [habitat with the raster being transformed to factor, Rhabitat <- as.factor(Rhabitat)] and others are numerical.

Below is my current code to get the response plot for GLM models only (n= 10):
GLM_CS_summer <- bm_PlotResponseCurves (bm.out = myBiomodModelOut_CS_summer,
                              models.chosen = get_built_models(myBiomodModelOut_CS_summer)[c(1,9,17,25,33,41,49,57,65,73)],
                              new.env = get_formal_data(myBiomodModelOut_CS_summer,'expl.var'),
                              show.variables= get_formal_data(myBiomodModelOut_CS_summer,'expl.var.names'),
                              do.bivariate = FALSE,
                              fixed.var = 'mean',
                              col = c("lightgrey","grey","darkgrey","lightblue","blue","darkblue","red","orange","yellow","lightgreen","darkgreen","black"),
                              main = "GLM",
                              ncol = 8, nrow = 1,
                              data_species = get_formal_data(myBiomodModelOut_CS_summer,'resp.var'))



Below is the message error I am having:
Using habitat as id variables
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?
geom_path: Each group consists of only one observation. Do you need to adjust the group aesthetic?

Have you been able to resolve this issue, assuming it is the same as Luis' one.

Many thanks in advance.

Delphine

**Comment by MayaGueguen** [comment]
Hello there,

I just did some corrections to the [bm_PlotResponseCurves](https://biomodhub.github.io/biomod2/reference/bm_PlotResponseCurves.html) function, that should enable it to "deal" with categorical variables with the following strategy : 
- categorical variables can be used _to plot 1D response curves_.
When not focal variable, the mode value is used ; when focal variable, the variable is transformed `as.numeric` for rendering plot.

- categorical variables are _still automatically removed when doing bivariate plot_.


Might still not be the optimum...
Maya

**Comment by DelphineChabanne** [comment]
Hi Maya,

Thank you for fixing the bm_PLotResponseCurves function. I see what it does and it looks good for the need. 

I am still having the same error though for the projection with my categorical variable. "variable 'habitat' was fitted with type "factor" but type "numeric" was supplied". Unfortunately, the results do change a bit if I were to do it with my categorical variable as a numerical one. 
Do you have any suggestions? 

Thank you for your help and your time.

Delphine

**Comment by rpatin** [comment]
Hello Delphine,

I suppose you were also using the CTA model for your projection (among other models). We just found that predicting CTA model with categorical variables on a raster was not supported by rpart package. Predicting CTA with categorical variables with a data.frame should be possible but is not yet possible I think.

I just did some correction that should warn you in such situation and switch off CTA raster prediction when using categorical variables, which should hopefully solve your problem. 

Rémi
